source/lcc/doc/bprint.1
.\" $Id: bprint.1 355 2007-02-18 22:08:49Z drh $
.TH BPRINT 1 "local \- $Date: 2007-02-19 09:08:49 +1100 (Mon, 19 Feb 2007) $"
.SH NAME
bprint \- expression profiler
.SH SYNOPSIS
.B bprint
[
.I option ...
]
[
.I file ...
]
.SH DESCRIPTION
.I bprint
produces on the standard output a listing of the programs compiled by
.I lcc
with the
.B \-b
option.
Executing an
.B a.out
so compiled appends profiling data to
.BR prof.out .
The first token of each expression in the listing is preceded
by the number of times it was executed
enclosed in angle brackets as determined from the data in
.BR prof.out .
.I bprint
interprets the following options.
.TP
.B \-c
Compress the
.B prof.out
file, which otherwise grows with every execution of
.BR a.out .
.TP
.B \-b
Print an annotated listing as described above.
.TP
.B \-n
Include line numbers in the listing.
.TP
.B \-f
Print only the number of invocations of each function.
A second
.B \-f
summarizes call sites instead of callers.
.TP
.BI \-I \*Sdir
specifies additional directories in which to seek
files given in
.B prof.out
that do not begin with `/'.
.PP
If any file names are given, only the requested data for those files are printed
in the order presented.
If no options are given,
.B \-b
is assumed.
.SH FILES
.PP
.ta \w'$LCCDIR/liblcc.{a,lib}XX'u
.nf
prof.out	profiling data
$LCCDIR/liblcc.{a,lib}	\fIlcc\fP-specific library
.SH "SEE ALSO"
.IR lcc (1), 
.IR prof (1)
.SH BUGS
Macros and comments can confuse
.I bprint
because it uses post-expansion source coordinates
to annotate pre-expansion source files.
If
.I bprint
sees that it's about to print a statement count
.I inside
a number or identifier, it moves the count to just
.I before
the token.
.PP
Can't cope with an ill-formed
.BR prof.out .


CHANGES.TXT
Catalina release 7.6.3 is a full release. 

WARNING: DO NOT INSTALL THIS RELEASE OVER A PRIOR VERSION OF CATALINA AND
DO NOT LINK OR COMBINE PROGRAMS COMPILED WITH THIS VERSION WITH CODE (E.G. 
OBJECT FILES OR LIBRARIES) COMPILED WITH VERSIONS OF CATALINA PRIOR TO 7.0.

Instead, either uninstall any previous Catalina release before installing, 
or else install this release to a different location.

The following sections list the changes that have been made in releases
7.0 onwards of Catalina. If you have not used a previous release of Catalina, 
you can skip the rest of this README and instead go straight to the Catalina 
tutorial documents.

RELEASE 7.6.3

New Functionality
-----------------

1. The Catalina Catapult utility now ignores pragmas it does not recognize,
   but leaves them intact. Only one warning message is now issued (for the 
   first such pragma) even if there are multiple unrecognised pragmas. This 
   allows primary and secondary programs to use the Catalina parallelizer.

2. Four new demo programs are now included (in demos\catapult) for the
   Propeller 1 and 2 to demonstrate the use of the Catalina parallelizer 
   with Catapult:

   For the Propeller 1:
      ll_p_p1.c - Demonstrates a parallelized primary program.
      ll_s_p1.c - Demonstrates a parallelized secondary program.

   For the Propeller 2:
      ll_p_p2.c - Demonstrates a parallelized primary and secondary program.
      ll_s_p2.c - Demonstrates two parallelized secondary programs.

3. The Catalina parallelizer utility now unregisters the factory kernel cogs 
   when they are stopped via the "#pragma propeller stop" pragma. Previously 
   these cogs were stopped but not unregistered.

4. The Catalina catapult utility now unregisters the secondary kernel cog 
   when a secondary program terminates by exiting its function. Previously, 
   these cogs were stopped but not unregistered.

5. On the Propeller 1, the cache cog is now registered when used. This means
   the registry is displayed correctly, and that this cog is correctly marked
   as being in use if a program wants to search the registry to locate an 
   unused cog (normally, ANY_COG is used to start a new cog, which does not
   depend on whether or not the cog was registered). This was already the 
   case on the Propeller 2.

Other Changes
-------------

1. The Catalina Geany command to build utilities did not specify that it
   should run in the project directory, not the current directory. The 
   version in this release fixes this for all new installations of Geany, 
   but it may not fix it for existing versions of Geany or existing projects.
 
   For existing versions of Geany, select the "Build -> Set Build Commands" 
   menu item, and add %p to the "Working Directory" field of the "Build 
   Utilities" command. 

   For existing project files, open the project and select the "Project -> 
   Properties" menu item, then select the "Build" tab and add %p to the 
   "Working Directory" field of the "Build Utilities" command.
 
   Affected the Propeller 1 and Propeller 2 on Windows and Linux.

2. Fixed a bug that may have led to some symbols being defined multiple
   times in complex parallelized or multi-model programs.

   The symbols that may have ended up multiply defined were:
      cmmd_array, lmmd_array, nmmd_array (on the Propeller 1 & 2), and
      CMM_LUT_LIBRARY_array, LMM_LUT_LIBRARY_array and NMM_LUT_LIBRARY_array
      (on the Propeller 2 only).

   Affected the Propeller 1 and Propeller 2 under Catalyst, Windows and Linux.

3. Some superseded, unused (and undocumented) functions have been removed 
   from the threads libraries:

      _threadstart_CMM() and _threadstart_CMM_cog()
      _threadstart_LMM() and _threadstart_LMM_cog()
      _threadstart_NMM() and _threadstart_NMM_cog()

   The following functions should be used instead:

      _threaded_cogstart_CMM_cog()
      _threaded_cogstart_LMM_cog()
      _threaded_cogstart_NMM_cog()

   Affected the Propeller 1 and Propeller 2 under Catalyst, Windows and Linux.

4. The example of using the Catalina Parallelizer with the Propeller version 
   of gcc was incorrectly modified to include Catalina's "prop2.h" rather 
   than gcc's "propeller2.h", which prevented it compiling.


RELEASE 7.6.2

New Functionality
-----------------

1. Two new catapult multi-model demo programs have been added, which
   demonstrate a threaded secondary program being executed from an XMM 
   primary program (which would not support multi-threading if it was 
   compiled as a simple XMM program). See demos\catapult\thread_p1.c and 
   demos\catapult\thread_p2.c - note that the 'build_utilities' script
   must be used to build an XMM loader suitable to load them.

2. The default pins used for the second serial port when the 2 port or 8 
   port serial plugins were in use (pins 50 & 52) conflicted with the pins 
   used for the PSRAM on the P2-EC32MB, and may conflict with the pins 
   used for the HyperFlash/HyperRAM add-on board, which made these plugins 
   unusable in XMM programs. The second port has now been disabled by 
   default by setting the tx and rx pins to -1 on all P2 platforms.

Other Changes
-------------

1. Fixed a problem with Catapult when a secondary program required threads
   (e.g. the secondary pragma included "options(-lthreads)") but the 
   primary program did not - the secondary program was being compiled to use 
   the threads library but was being started using the non-threaded kernel
   and thus would not execute correctly. Affected the Propeller 1 and 
   Propeller 2 under Windows and Linux.

2. Fixed a bug in the XMM dynamic kernel, which broke the start.c program in
   the demos\p2_psram folder, which needs to load the XMM kernel dynamically.
   This was the only program that uses that particular version of the kernel. 
   Also, the Makefile referred to the target directory using a relative 

source/lcc/etc/bprint.c
#include "profio.c"
#include <assert.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* bprint [ -c | -Idir... | -f | -b | -n ] [ file... ]
 * annotate listings of files with prof.out data
 */

static char rcsid2[] = "$Id: bprint.c 355 2007-02-18 22:08:49Z drh $";

#define NDIRS (sizeof dirs/sizeof dirs[0] - 1)
#define NELEMS(a) ((int)(sizeof (a)/sizeof ((a)[0])))

char *progname;
int number;
char *dirs[20];
int fcount;

void *alloc(unsigned);
void emitdata(char *);
void printfile(struct file *, int);
void printfuncs(struct file *, int);

void *allocate(unsigned long n, unsigned a) { return alloc(n); }

void *newarray(unsigned long m, unsigned long n, unsigned a) {
	return alloc(m*n);
}

int main(int argc, char *argv[]) {
	int i;
	struct file *p;
	void (*f)(struct file *, int) = printfile;

	progname = argv[0];
	if ((i = process("prof.out")) <= 0) {
		fprintf(stderr, "%s: can't %s `%s'\n", progname,
			i == 0 ? "open" : "interpret", "prof.out");
		exit(1);
	}
	for (i = 1; i < argc && *argv[i] == '-'; i++)
		if (strcmp(argv[i], "-c") == 0) {
			emitdata("prof.out"); 
			exit(0);
		} else if (strcmp(argv[i], "-b") == 0)
			f = printfile;
		else if (strcmp(argv[i], "-f") == 0) {
			fcount++;
			f = printfuncs;
		} else if (strcmp(argv[i], "-n") == 0)
			number++;
		else if (strncmp(argv[i], "-I", 2) == 0) {
			int j;
			for (j = 0; j < NDIRS && dirs[j]; j++)
				;
			if (j < NDIRS)
				dirs[j] = &argv[i][2];
			else
				fprintf(stderr, "%s: too many -I options\n", progname);
		} else {
			fprintf(stderr, "usage: %s [ -c | -b | -n | -f | -Idir... ] [ file... ]\n", progname);
			exit(1);
		}
	for (p = filelist; p; p = p->link)
		qsort(p->counts, p->count, sizeof *p->counts, compare);
	if (i < argc) {
		int nf = i < argc - 1 ? 1 : 0;
		for ( ; i < argc; i++, nf ? nf++ : 0)
			if (p = findfile(string(argv[i])))
				(*f)(p, nf);
			else
				fprintf(stderr, "%s: no data for `%s'\n", progname, argv[i]);
	} else {
		int nf = filelist && filelist->link ? 1 : 0;
		for (p = filelist; p; p = p->link, nf ? nf++ : 0)
			(*f)(p, nf);
	}
	return 0;
}

/* alloc - allocate n bytes or die */
void *alloc(unsigned n) {
	void *new = malloc(n);

	assert(new);
	return new;
}

/* emitdata - write prof.out data to file */
void emitdata(char *file) {
	FILE *fp;

	if (fp = fopen(file, "w")) {
		struct file *p;
		for (p = filelist; p; p = p->link) {
			int i;
			struct func *q;
			struct caller *r;
			fprintf(fp, "1\n%s\n", p->name);
			for (i = 0, q = p->funcs; q; i++, q = q->link)
				if (r = q->callers)
					for (i--; r; r = r->link)
						i++;
			fprintf(fp, "%d\n", i);
			for (q = p->funcs; q; q = q->link)
				if (q->count.count == 0 || !q->callers)
					fprintf(fp, "%s 1 %d %d %d ? ? 0 0\n", q->name, q->count.x,
						q->count.y, q->count.count);
				else
					for (r = q->callers; r; r = r->link)
						fprintf(fp, "%s 1 %d %d %d %s %s %d %d\n", q->name, q->count.x,
							q->count.y, r->count, r->name, r->file, r->x, r->y);
			fprintf(fp, "%d\n", p->count);
			for (i = 0; i < p->count; i++)
				fprintf(fp, "1 %d %d %d\n", p->counts[i].x,
					p->counts[i].y, p->counts[i].count);
		}
		fclose(fp);
	} else
		fprintf(stderr, "%s: can't create `%s'\n", progname, file);
}

/* openfile - open name for reading, searching -I directories */
FILE *openfile(char *name) {
	int i;
	FILE *fp;

	if (*name != '/')	
		for (i = 0; dirs[i]; i++) {
			char buf[200];
			sprintf(buf, "%s/%s", dirs[i], name);
			if (fp = fopen(buf, "r"))
				return fp;
		}
	return fopen(name, "r");
}

/* printfile - print annotated listing for p */
void printfile(struct file *p, int nf) {
	int lineno;
	FILE *fp;
	char *s, buf[512];
	struct count *u = p->counts, *r, *uend;

	if (u == 0 || p->count <= 0)
		return;
	uend = &p->counts[p->count];
	if ((fp = openfile(p->name)) == NULL) {
		fprintf(stderr, "%s: can't open `%s'\n", progname, p->name);
		return;
	}
	if (nf)
		printf("%s%s:\n\n", nf == 1 ? "" : "\f", p->name);
	for (lineno = 1; fgets(buf, sizeof buf, fp); lineno++) {
		if (number)
			printf("%d\t", lineno);
		while (u < uend && u->y < lineno)
			u++;
		for (s = buf; *s; ) {
			char *t = s + 1;
			while (u < uend && u->y == lineno && u->x < s - buf)
				u++;
			if (isalnum(*s) || *s == '_')
				while (isalnum(*t) || *t == '_')
					t++;
			while (u < uend && u->y == lineno && u->x < t - buf) {
				printf("<%d>", u->count);
				for (r = u++; u < uend && u->x == r->x && u->y == r->y && u->count == r->count; u++)
					;
			}
			while (s < t)
				putchar(*s++);
		}
		if (*s)
			printf("%s", s);
	}
	fclose(fp);
}

/* printfuncs - summarize data for functions in p */
void printfuncs(struct file *p, int nf) {
	struct func *q;

	if (nf)
		printf("%s:\n", p->name);
	for (q = p->funcs; q; q = q->link)
		if (fcount <= 1 || q->count.count == 0 || !q->callers)
			printf("%d\t%s\n", q->count.count, q->name);
		else {
			struct caller *r;
			for (r = q->callers; r; r = r->link)
				printf("%d\t%s\tfrom %s\tin %s:%d.%d\n", r->count, q->name, r->name,
					r->file, r->y, r->x + 1);
		}
		
}

/* string - save a copy of str, if necessary */
char *string(const char *str) {
	static struct string { struct string *link; char str[1]; } *list;
	struct string *p;

	for (p = list; p; p = p->link)
		if (strcmp(p->str, str) == 0)
			return p->str;
	p = alloc(strlen(str) + sizeof *p);
	strcpy(p->str, str);
	p->link = list;
	list = p;
	return p->str;
}
